solr/core/src/java/org/apache/solr/schema/BinaryField.java

```diff
@@ -112,7 +125,31 @@ public IndexableField createField(SchemaField field, Object val) {
       len = buf.length;
     }
 
-    return new org.apache.lucene.document.StoredField(field.getName(), buf, offset, len);
+    return new BytesRef(buf, offset, len);
+  }
+
+  @Override
+  public List<IndexableField> createFields(SchemaField field, Object val) {
+    IndexableField fval = createField(field, val);
+
+    if (field.hasDocValues() && !field.multiValued()) {
+      IndexableField docval = new BinaryDocValuesField(field.getName(), getBytesRef(val));
```

Comment:
This is the crux of the changes - we want to use Lucene `BinaryDocValuesField` if user specified `docValues="true"` in the schema.